### ISSUE TITLE
CLI: Setting sourceRoot option removes folder structure

### ISSUE THREAD

**Issue by cr0** [jashkenas/coffeescript#3075]
Hi there

is it on purpose that setting the `sourceRoot` option removes the coffee files' folder structure from the generated source map?

With sourceRoot set to `coffee/`

```
./assets/coffee/lib/utils.coffee
```

the generated source map at `./public/js/lib/utils.map` will be:

```
"sourceRoot": "/coffee/",
"sources": [
     "utils.coffee"
]
```

This makes Chrome unable to find the attached `coffee` file for a `javascript` file. As you can see on the attached file the old folder structure is completely removed. 
![flat-structure](https://f.cloud.github.com/assets/854965/826895/792f38fe-f091-11e2-99f0-bdd3678a249e.png)
So Chrome reconstructs the folder structure based on the source maps wrong. This makes it impossible for me to debug things because every request for a coffee file in a subfolder ends in a 404.

I'm using grunt to build:

```
coffee:
  client:
    options:
      sourceMap: true
      sourceRoot: '/coffee/'
    files: [
      expand:   true
      cwd:      'assets/coffee/'
      src:      ['**/*.coffee']
      dest:     'public/js/'
      ext:      '.js'
    ]
```

Is something wrong with my build setup? Or is this a bug (or a feature ;))? I'm on 1.6.3.

**Comment by GeoffreyBooth** [comment]
Is this still an issue?

**Comment by GeoffreyBooth** [comment]
So I looked into this a little. Using the folder structure from #3296, I wrote a script to compile that `.coffee` file (because `sourceRoot` is only an API option, not a command-line option):

```coffee
fs = require 'fs'
CoffeeScript = require 'coffeescript'

filename = './website_root/coffee/work.coffee'

{ js, v3SourceMap, sourceMap } = CoffeeScript.compile fs.readFileSync(filename, 'utf-8'),
	filename: filename
	bare: yes
	sourceMap: yes

console.log v3SourceMap
```
And it printed this (2.0.0-beta4):
```json
{
  "version": 3,
  "file": "",
  "sourceRoot": "",
  "sources": [
    ""
  ],
  "names": [],
  "mappings": "AAAA,OAAO,CAAC,GAAR,CAAY,IAAI,CAAC,GAAL,CAAA,CAAZ"
}
```
Well well well, not at all what we want! Empty `sources` and `sourceRoot`?

Looking into `sourcemap.litcoffee`, I found this:

```litcoffee
Produce the canonical JSON object format for a "v3" source map.

        v3 =
          version:    3
          file:       options.generatedFile or ''
          sourceRoot: options.sourceRoot or ''
          sources:    options.sourceFiles or ['']
          names:      []
          mappings:   buffer
```

So apparently, if you pass the undocumented `sourceRoot` and `sourceFiles` in as options, they get output as is into the source map. And indeed, changing my script:

```coffee
{ js, v3SourceMap, sourceMap } = CoffeeScript.compile fs.readFileSync(filename, 'utf-8'),
	filename: filename
	bare: yes
	sourceMap: yes
	sourceRoot: './website_root/coffee/'
	sourceFiles: [filename]
```
produced:
```json
{
  "version": 3,
  "file": "",
  "sourceRoot": "./website_root/coffee/",
  "sources": [
    "./website_root/coffee/work.coffee"
  ],
  "names": [],
  "mappings": "AAAA,OAAO,CAAC,GAAR,CAAY,IAAI,CAAC,GAAL,CAAA,CAAZ"
}
```

So the original bug above is no longer reproducible, or perhaps was a bug in Grunt or `grunt-coffee`. So we can close this issue.

But as for the empty strings, we have options:

1. Leave the code alone and just document `sourceRoot` and `sourceFiles` in the [Node API docs](http://coffeescript.org/v2/#nodejs-usage).
2. Document `sourceRoot`, but generate `sources` from `[filename]` (since the `compile` method takes a string as input, so by definition you’re only compiling one file) if `sourceFiles` isn’t specified but `filename` is.

Also should the default values for these really be `null`? @lydell, I see you have some history with this: https://github.com/mozilla/source-map/issues/123

**Comment by GeoffreyBooth** [comment]
Fixed via #4671.
